ql/src/java/org/apache/hadoop/hive/ql/exec/FileSinkOperator.java

```diff
@@ -1063,7 +1076,11 @@ public void process(Object row, int tag) throws HiveException {
       // RecordUpdater expects to get the actual row, not a serialized version of it.  Thus we
       // pass the row rather than recordValue.
       if (conf.getWriteType() == AcidUtils.Operation.NOT_ACID || conf.isMmTable() || conf.isCompactionTable()) {
-        rowOutWriters[findWriterOffset(row)].write(recordValue);
+        writerOffset = bucketId;
+        if (!conf.isCompactionTable()) {
+          writerOffset = findWriterOffset(row);
+        }
+        rowOutWriters[writerOffset].write(recordValue);
```

Comment:
Are the results in order? Can we close writers for old buckets? That could save memory

Reply:
They should be in order, because the result temp table for the compaction is created like "clustered by (`bucket`) sorted by (`bucket`, `originalTransaction`, `rowId`) into 10 buckets". I would assume that because of this, the rows in the table will be in order by bucket, originalTransaction and rowId. I haven't seen otherwise during my testing.
I don't think we can close the writers here, because they will be used in the closeOp method as well and they are closed there.